docs/developers/decisions/backend/0018-durable-jobs.md

```diff
@@ -0,0 +1,238 @@
+can use all the normal C# features like `async` and `await` and `foreach` loops:
+
+
+```csharp
+
+    public class WaitMany : AOrchestration<WaitMany, int, int[]>
+    {
+        protected override async Task<int> Run(Context context, int[] inputs)
+        {
+            // Add all the tasks to a list
+            var tasks = new List<Task<int>>();
+            foreach (var input in inputs)
+            {
+                tasks.Add(SquareJob.RunSubJob(context, input));
+            }
+    
+            // Wait for all the tasks to complete
+            await Task.WhenAll(tasks);
+    
+            return tasks.Select(t => t.Result).Sum();
+        }
+    }
+    
+    public class AsyncLinqJob : AOrchestration<AsyncLinqJob, int, int[]>
+    {
+        protected override async Task<int> Run(Context context, int[] ints)
+        {
+            var sum = 0;
+            
+            // Even Async Linq works
+            await foreach (var val in ints.SelectAsync(async x => await SquareJob.RunSubJob(context, x)))
+            {
+                sum += val;
+            }
+            return sum;
+        }
+    }
+
+```
+
+### Implementation Details
+
+Internally each job becomes an actor (a Task and a queue of messages). Jobs communicate to eachother via messages, and the 
+changes and updates of a single job are single threaded, this is done mostly to contain the complexity of the implementation, and to 
+allow the use of standard mutable collections in the job code.
+
+The concept of "Replaying" is not very intuitive, but it is easy to understand: 
+
+* When a job is run, it is handed a `Context` object in this object is
+  * `HistoryIndex` - a stack pointer of sorts into the history of the job, it starts at 0 
+  * `List<HistoryEntry>` - this is the history of the job
+    * `HistoryEntry` is a record of the following values
+      * `ChildJobId` - The id (a guid) of the child job that was run
+      * `Status` - Either `Running`, `Completed`, or `Failed` depending on the state of the sub-job
+      * `Job` the singleton instance of the sub-job that was run (for serialization purposes)
+      * `Result` - The result of the sub-job if it was completed, this will be an error message if the job failed
+* When a call is made to `RunSubJob` the job system will look at the history index
+ * If the index is higher than the length of the history:
+   * A sub job will be created and started, it's JobId will be added to the history
+   * The current history index will be handed to the job as the "parent history index"
+ * If the index is less than the length of the history:
+   * This means we're replaying the job and we've already spawned this sub-job
+   * So we look at the status of the sub-job
+     * If it's `Running` we return a `Task.FromException(new WaitException())` which will cause the current replay
+ event to abort and the job will be considered to be paused
```

Comment:
What are the reasons for throwing the exception in case the status of the SubJob is `Running`?
My first instinct would be to issue a Cancellation to the SubJob and restart it if it cancels within a certain timeout, throw exception otherwise. 
I'm guessing this wouldn't be possible?

Reply:
That might be possible, but it does mean that a situation could occur where multiple sub-jobs complete at once and then their results would have to be coordinated as to not restart the execution too many times. Mostly it's just a simpler model to understand this way